feat(actions): getActionPath() (#12721)
withastro/astro: .changeset/wise-boxes-develop.md

---
'astro': minor
---

Exports a new `getActionPath()` helper from `astro:actions`

In most cases, calling an action as `actions.like()` is enough. But sometimes it's not enough, for example:

- You want to pass custom headers
- You want to call the actions endpoint without `fetch`, eg. using the `navigator.sendBeacon` API

> That's why you can now use the `getActionPath()` helper. Pass an action to it to get the pathname (prefixed by your `based` configuration) to the action:

Reviewer: ```suggestion
That's why you can now use the `getActionPath()` helper. Pass an action to it to get the pathname to the action (prefixed by your configured `base` value) :
```

---
'astro': minor
---

Exports a new `getActionPath()` helper from `astro:actions`

> The `getActionPath()` utility accepts an action and returns a URL path so you can execute an action call as a `fetch()` operation directly. This allows you to provide details such as custom headers when you call your action.

Reviewer 2: ```suggestion
Astro 5.1 introduces a new helper function, `getActionPath()` to give you more flexibility when calling your action.

Calling `getActionPath()` with your action returns its URL path so you can make a `fetch()` request with custom headers, or use your action with an API such as `navigator.sendBeacon()`.
```

> This example shows how to call a defined `like` action passing the `Authorization` header and the [`keepalive`](https://developer.mozilla.org/en-US/docs/Web/API/Request/keepalive) option:

Reviewer 2: ```suggestion
This example adds authorization to a `like` action by passing the `Authorization` header and the [`keepalive`](https://developer.mozilla.org/en-US/docs/Web/API/Request/keepalive) option:
```
Would that be an OK way to say this? Since this is not docs, it doesn't need to be *quite* as clinical. (But probably not as casual as the blog post.)

Author: I don't think it's 100% correct. You don't add authorization to the action, it assumes it's already setup and you just provide some credentials basically

Reviewer 2: OK! If you can come up with something in plain English kind of like: "This example passes credentials to an action requiring authentication, as well as an option to keep the associated request alive even if the user navigates away from the page:" then go for it! Otherwise, what you have is fine, and non-blocking!

```astro
<script>
// src/components/my-component.astro
import { actions, getActionPath } from 'astro:actions'

await fetch(getActionPath(actions.like), {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    Authorization: 'Bearer YOUR_TOKEN'
  },
  body: JSON.stringify({ id: 'YOUR_ID' }),
  keepalive: true

> })

Reviewer 3: Thanks for adding this! I think it's definitely worth noting that **actions do not return plain JSON.** They return values that are parsable by the `devalue` library. I suggest using the `deserializeActionResult()` utility from `astro:actions` for this.

Author: Ah good call thank you!

Reviewer 2: Thanks @bholmesdev -- is *showing* using `deserializeActionResult()` in this context concise? I realized we only have a short API reference sentence for this (no example usage), and no example using this anywhere in docs. (So, I'm not sure if this can be shown quickly or not.)

This would be a great addition to the actual docs (where I don't really care if it's concise or not), and if it's easy to do here, would also be a great place to put it!

---
'astro': minor
---

Adds a new `getActionPath()` helper available from `astro:actions`

Astro 5.1 introduces a new helper function, `getActionPath()` to give you more flexibility when calling your action.

> Calling `getActionPath()` with your action returns its URL path so you can make a `fetch()` request with custom headers, or use your action with an API such as `navigator.sendBeacon()`.

Reviewer 2: ```suggestion
Calling `getActionPath()` with your action returns its URL path so you can make a `fetch()` request with custom headers, or use your action with an API such as `navigator.sendBeacon()`. Then, you can [handle the custom-formatted returned data](https://docs.astro.build/en/guides/actions/#handling-returned-data) as needed, just as if you had called an action directly.
```
@florian-lefebvre @bholmesdev Do you think this addresses handling the returned data concern OK (enough for the changeset?)
